File: 202305/1406.py
class Solution:
    def solve2(self, a, k, n):
        if k >= n:
            return 0
        if self.dp2[k] != -1:
            return self.dp2[k]

        ans = a[k] - self.solve(a, k + 1, n)
        curr = a[k]
        for i in range(k + 1, min(n, k + 3)):
            curr += a[i]
            ans = max(ans, curr - self.solve(a, i + 1, n))

        self.dp2[k] = ans
        return ans

    def solve(self, a, k, n):
        if k >= n:
            return 0
        if self.dp[k] != -1:
            return self.dp[k]

        ans = a[k] - self.solve2(a, k + 1, n)
        curr = a[k]
        for i in range(k + 1, min(n, k + 3)):
            curr += a[i]
            ans = max(ans, curr - self.solve2(a, i + 1, n))

        self.dp[k] = ans
        return ans

    def stoneGameIII(self, stoneValue: List[int]) -> str:
        n = len(stoneValue)
        self.dp = [-1] * n
        self.dp2 = [-1] * n
        s = sum(stoneValue)
        ans = self.solve(stoneValue, 0, n)
        print(ans)

        alice = (s + ans) / 2
        bob = (s - ans) / 2
        if alice > bob:
            return "Alice"
        elif alice < bob:
            return "Bob"

        return "Tie"

File: 202608/1406.cpp
class Solution {
  const int oo = 1e9;
  vector<int> stone;
  int n;
  int dp[2][50001];

  int f(int i, int round) {
    if (i >= stone.size())
      return 0;
    if (dp[round][i] != -1)
      return dp[round][i];
    int &ans = dp[round][i];
    ans = -oo;
    int tt = 0;
    for (int j = i; j < min(i + 3, n); ++j) {
      tt += stone[j];
      ans = max(ans, tt - f(j + 1, 1 - round));
    }
    return ans;
  };

public:
  string stoneGameIII(vector<int> &stoneValue) {

    stone = stoneValue;
    n = stone.size();
    memset(dp, -1, sizeof(dp));
    int diff = f(0, 0);
    if (diff > 0)
      return "Alice";
    if (diff < 0)
      return "Bob";
    return "Tie";
  }
};
